From 2df80d741090665c9ec4da8884c9098387c25167 Mon Sep 17 00:00:00 2001
From: Frank Wunderlich <frank-w@public-files.de>
Date: Sat, 17 Jan 2026 12:38:40 +0100
Subject: [PATCH v6 1/3] cmd: mem: add command for getting ram size for use in
 scripts

Add a command for getting detected ram size with possibility to
assign it to an environment variable.

example usage:

BPI-R4> memsize
4096 MiB
BPI-R4> memsize memsz
BPI-R4> printenv memsz
memsz=4096
BPI-R4>

board with 8GB ram:

BPI-R4> memsize
8192 MiB
BPI-R4> memsize memsz
BPI-R4> printenv memsz
memsz=8192
BPI-R4>

Signed-off-by: Frank Wunderlich <frank-w@public-files.de>
---
v6: add missing return for error in env_set_ulong
v5: move msize to meminfo and drop first param (always display as MiB)
    rename msize to memsize
v4: drop rounding to full MB/GB as it leads to wrong display
v3: add missing ifdefs
v2: add Kconfig entry
---
 cmd/Kconfig   |  6 ++++++
 cmd/meminfo.c | 25 +++++++++++++++++++++++++
 2 files changed, 31 insertions(+)

--- a/cmd/Kconfig
+++ b/cmd/Kconfig
@@ -931,6 +931,12 @@ config CMD_MEMINFO_MAP
 
 	  See doc/usage/cmd/meminfo.rst for more information.
 
+config CMD_MEMSIZE
+	bool "memsize"
+	depends on CMD_MEMINFO
+	help
+	  Get RAM via command for use in scripts.
+
 config CMD_MEMORY
 	bool "md, mm, nm, mw, cp, cmp, base, loop"
 	default y
--- a/cmd/meminfo.c
+++ b/cmd/meminfo.c
@@ -8,10 +8,12 @@
 #include <bootstage.h>
 #include <command.h>
 #include <display_options.h>
+#include <env.h>
 #include <lmb.h>
 #include <malloc.h>
 #include <mapmem.h>
 #include <asm/global_data.h>
+#include <linux/sizes.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -98,8 +100,31 @@ static int do_meminfo(struct cmd_tbl *cm
 	return 0;
 }
 
+#ifdef CONFIG_CMD_MEMSIZE
+static int do_mem_size(struct cmd_tbl *cmdtp, int flag, int argc,
+		       char *const argv[])
+{
+	u64 memsize = gd->ram_size / SZ_1M;
+
+	if (argc > 1)
+		return env_set_ulong(argv[1], memsize);
+	else
+		printf("%lld MiB\n", memsize);
+
+	return 0;
+}
+#endif /* CONFIG_CMD_MEMSIZE */
+
 U_BOOT_CMD(
 	meminfo,	1,	1,	do_meminfo,
 	"display memory information",
 	""
 );
+
+#ifdef CONFIG_CMD_MEMSIZE
+U_BOOT_CMD(
+	memsize,	2,	1,	do_mem_size,
+	"get detected ram size in MiB, optional set env variable with value",
+	"[envvar]"
+);
+#endif /* CONFIG_CMD_MEMSIZE */
